lib/bricks-net-api-types.js

```diff
@@ -0,0 +1,1176 @@
+'use strict';
+
+var fs = require('fs');
+var util = require('util');
+var inherits = require('util').inherits;
+var net = require('net');
+
+var when = require('when');
+var whenPipeline = require('when/pipeline');
+
+var cppArguments = require('./cpp-arguments');
+var promiseHelpers = require('./promise-helpers');
+
+var HTTPResponseCode = require('./bricks-net-http-codes').HTTPResponseCode;
+
+var BRICKS_THROW = require('./bricks-throw');
+var HTTPNoBodyProvidedException = require('./bricks-net-exceptions').HTTPNoBodyProvidedException;
+var HTTPRedirectNotAllowedException = require('./bricks-net-exceptions').HTTPRedirectNotAllowedException;
+var HTTPRedirectLoopException = require('./bricks-net-exceptions').HTTPRedirectLoopException;
+var ConnectionResetByPeer = require('./bricks-net-exceptions').ConnectionResetByPeer;
+
+var URL = require('./bricks-net-url').URL;
+
+
+function virtualMethod() {
+	/* istanbul ignore next */
+	return function () {
+		throw new Error('VIRTUAL METHOD');
+	};
+}
+function notImplementedMethod() {
+	/* istanbul ignore next */
+	return function () {
+		throw new Error('NOT IMPLEMENTED METHOD');
+	};
+}
+
+
+// Structures to define HTTP requests.
+// Support GET and POST.
+// The syntax for creating an instance of a GET request is GET is `GET(url)`.
+// The syntax for creating an instance of a POST request is POST is `POST(url, data, content_type)`'.
+// Alternatively, `POSTFromFile(url, file_name, content_type)` is supported.
+// Both GET and two forms of POST allow `.UserAgent(custom_user_agent)`.
+function HTTPRequestBase(url) {
+	cppArguments.assert('HTTPRequestBase', [
+		[ cppArguments.assertion('string', 'const std::string&', 'url') ]
+	], arguments);
+	
+	this.url = url;
+	this.custom_user_agent = "";
+	this.allow_redirects = false;
+}
+HTTPRequestBase.prototype.UserAgent = function (new_custom_user_agent) {
+	cppArguments.assert('HTTPRequestBase::UserAgent', [
+		[ cppArguments.assertion('string', 'const std::string&', 'new_custom_user_agent') ]
+	], arguments);
+	
+	this.custom_user_agent = new_custom_user_agent;
+	
+	return this;
+};
+HTTPRequestBase.prototype.AllowRedirects = function (allow_redirects_setting) {
+	cppArguments.assert('HTTPRequestBase::AllowRedirects', [
+		[ cppArguments.assertion('bool', 'bool', 'allow_redirects_setting', cppArguments.ASSERTION_MODE_OPTIONAL) ]
+	], arguments);
+	
+	if (typeof allow_redirects_setting === 'undefined') {
+		allow_redirects_setting = true;
+	}
+	
+	this.allow_redirects = allow_redirects_setting;
+	
+	return this;
+};
+
+
+function GET(url) {
+	var _this = this;
+	
+	if (!(_this instanceof GET)) {
+		_this = new GET(url);
+		return _this;
+	}
+	
+	cppArguments.assert('GET', [
+		[
+			cppArguments.assertion('string', 'const std::string&', 'url'),
+			function (url) {
+				HTTPRequestBase.call(_this, url);
+			}
+		]
+	], arguments);
+}
+inherits(GET, HTTPRequestBase);
+
+
+function POST(url) {
+	var _this = this;
+	
+	if (!(_this instanceof POST)) {
```

Comment:
I could try `Object.create` and then call the constructor function on the created object instead of the `new` operator. This would allow to use `apply` for the call. Thanks Node.js has a modern JS engine.